### DISCUSSION TITLE
Composite out...

### DISCUSSION THREAD

**Discussion by fdivitto** [fdivitto/FabGL#192] [Show and tell]
Work in progress... SCH phase not so bad, Bruch sequence OK. There is a lot of work still to do...

BTW: it could be useful to have a Tektronix VM700 with NTSC option, I have only the PAL version.

![IMG_6175](https://user-images.githubusercontent.com/9754506/145706771-ca811f7a-9072-4a19-a182-fdc66de29f90.JPG)
![IMG_6174](https://user-images.githubusercontent.com/9754506/145706774-c6899adb-2a76-4499-a332-45fb54e00020.JPG)
![IMG_6173](https://user-images.githubusercontent.com/9754506/145706776-1ce76c93-8eb4-43fd-aa77-5082b6d90467.JPG)

**Comment by fdivitto** [comment]
slow progresses, anyway now we have almost perfect SCH phase (8 fields PAL Bruch sequence)!

![IMG_6255](https://user-images.githubusercontent.com/9754506/151032542-f58153c2-55da-4c28-a209-946cdcc3be6a.JPG)
![IMG_6256](https://user-images.githubusercontent.com/9754506/151032555-03bc228c-0cdd-48d7-bbb3-d558af9fd7fd.JPG)

**Comment by fdivitto** [comment]
**Habemus colores!!**   (...and source code this night)

PAL:
![IMG_6286](https://user-images.githubusercontent.com/9754506/154317991-42d3274e-aafc-4dc9-92cb-62eabb01ab50.JPG)

NTSC:
![IMG_6285](https://user-images.githubusercontent.com/9754506/154318026-0526ffd9-374d-4a0d-811a-92b9ff63321d.JPG)

**Reply by netpipe** [reply]
awesome, I look forward to using GL with this :) that tektronix looks like a great investment!

**Reply by netpipe** [reply]
https://www.ebay.ca/itm/202641744487 https://www.ebay.ca/itm/403459849077 I have both of these to test with.

**Reply by fdivitto** [reply]
I don't know, I'm writing just the software. Hardware implementations may use different connectors. At the moment I'm just using a baseline ESP32 board with gpio 25 directly connected to a composite monitor.

**Reply by ghost** [reply]
Looking good!  Well done.  Very exciting.

If you're designing new boards for TV out, it might be worth using the TRRS port like the Raspberry Pi does - integrating audio and composite out on the same connector.   

![Model-B-Plus-Audio-Video-Jack-Diagram](https://user-images.githubusercontent.com/22624099/154344321-b0cee82a-d0c1-4ac8-b698-0cff4cd09703.png)

Also you might be interested you can output VGA via USB mini B port - it's very compact and there's lots of cables available.....

![1-5M-Mini-USB-Male-5-pin-To-VGA-D-SUB-15-pins-Male-Adapter-Cable copy](https://user-images.githubusercontent.com/22624099/154344456-98aa4dc2-6dfc-4f0b-8d60-ab08d522e85a.jpg)

**Reply by netpipe** [reply]
on the ttygo boards  the exposed pins are 3v gnd io39 io34 - second row of pins io12 io14 io13 and io2. I hadnt gotten a chance to get on of the official boards yet but probably need one to do the pcemulator part better with 16mb.

**Reply by netpipe** [reply]
hopefully this runs on other arduino boards too like due, and mega.

**Reply by fdivitto** [reply]
> hopefully this runs on other arduino boards too like due, and mega.

This is a ESP32 specific project.
Also, composite out needs a DAC, and this is available just on GPIO 25 and 26.

**Reply by netpipe** [reply]
we can hope for better arduino's in the future where this should still be compatible.. with advancements in risc and arm. the current audio pins have DAC aswell ?

**Comment by netpipe** [comment]
do you have benchmarks for composite video vs vga rendering ?

**Comment by josephresearcher** [comment]
Dear Fabrizio
I'm glad you finally did it in such a short time
Proteus software does not have a library for CRT TV simulation for composite video and RGB inputs
I hope friends can design for him later

**Comment by fdivitto** [comment]
Now the repo contains Composite/BaseExample demo, that shows composite color video output (PAL and NTSC). Sample video here:

[Composite Output Sample](https://youtu.be/5QteUItkyVs)

I am still not satisfied with the quality of the color output, but I am actively working on it.

**Comment by ghost** [comment]
Looking great!

Are you planning for the VT100 to work on PAL/NTSC?

**Reply by fdivitto** [reply]
because display controllers in FabGL are replaceable, you should be able to do it right now (ie replace VGA16Controller with CVBS16Controller) . I doubt the quality is good enough to have 80 color columns anyway.

**Comment by ghost** [comment]
Hi Fabrizio - what are your plans for TV out from here?

I'd love to see an extended YouTube demo of as much as you can show running on TV please. Maybe VT100 in 40 column mode......

Can a single esp32 board have connectors for both VGA and composite out?

It would be great to see a board that had a combined sound/composite out port like on the Raspberry Pi with the TRRS port.

**Reply by fdivitto** [reply]
>>what are your plans for TV out from here?

I don't know, really! ;-)   

>Maybe VT100 in 40 column mode......

Could a 40 column terminal be really useful?

>Can a single esp32 board have connectors for both VGA and composite out?

Theoretically yes, also because just one pin is enough for the composite. Obviously not working at the same time.

**Comment by ghost** [comment]
>> Could a 40 column terminal be really useful?

I love old televisions.  I have quite a few.

Imagine this:
1: FabGL running a terminal in 40 column mode.
2: FabGL automatically telnets into a specific server on startup and runs a program.

At this point I have all the power of the server operating system to write text mode applications running on a TV.

Here's a fun idea:
1: FabGL displays a QR code on the 40 column TV screen/VT100 terminal.
2: Scan the QR code with your phone, which goes to a link .local multicast DNS hostname which sends keyboard control and touch gesture control back to FabGL via wifi or bluetooth, so your phone becomes the control device for FabGL!

Or image running MIcropython on FabGL on an old TV - I'd be able to build applications for all sorts of things, with output displayed on the TV.

Another idea: Teletext/Minitel is well suited to 40 column mode on a TV, maybe some sort of fun Teletext throwback client/server could be made.

I'd love it if FabGL could incorporate any of the ideas above, no doubt you are busy enough though!

I'm hoping LilyGo will make a 4MB esp32 with just a TRRS port like the Raspberry Pi for displaying audio/video and running FabGL.  Then I could buy one for each of my old TV's because it should be dirt cheap to make such an esp32.

**Comment by Zhornic** [comment]
Hey!
I checked the quality of the generated image on a large TV. Image quality in NTSC is noticeably better than in PAL. The contours of the elements almost do not blur and the fill is quite stable. On a large TV, the flaws are clearly visible and conspicuous even in NTSC. But if you display the image on the screen of a car radio with a diagonal of 7 inches, then the picture is formed already of a fairly decent quality. I'm waiting for the final version of the composite signal generation code.

I went through many options for the formation of a composite signal using the ESP32 and this program is the best in quality. But while I use a ready-made VGA-RCA converter, its quality is almost perfect.

**Reply by ghost** [reply]
Any chance you could put some of your testing onto YouTube?

**Reply by Zhornic** [reply]
https://www.drive2.ru/b/561070064489463999/
Description of my experiments in Russian. There are photos and videos.
Use electronic translators.

The beauty of the iron converter is that it produces a signal that fully complies with the specifications of the NTSC and the PAL. Therefore, this signal is understood by all TVs without exception. The signal generated by microcontrollers is often not perceived by TVs correctly and the image is either not at all, or it is distorted. Also, the converter allows you to accurately fit the image on the TV screen, shifting it vertically and horizontally, or changing its size and proportions.

I displayed a picture from a computer in a resolution of 1280x1024 and small letters were not readable on the TV, but in general the picture was very clear and stable. There was no flickering or blurring. Primitive images of 640*480 pixels will be in perfect quality.

**Reply by ghost** [reply]
Interesting blog post.  

This project also generates NTSC and PAL https://github.com/rossumur/esp_8_bit

Also this https://github.com/bitluni/ESP32CompositeVideo

There is more listed here: https://www.makeuseof.com/esp32-composite-video-projects/

Also, all Raspberry Pi's generate composite NTSC and PAL.

**Reply by Zhornic** [reply]
I tested all these projects on my esp32. The picture quality was very poor. Even on a small screen, it would look very bad. NTSC mode by FABGL provides sufficient quality for a small TV screen. I hope that the final version will generate an even higher quality signal.

**Reply by ghost** [reply]
Sounds exciting. Hopefully Fabrizio will make some cool demos to show off the TV composite out capabilities.

**Comment by exdiogene** [comment]
I am also interested to know the answer to "netpipe" question :

"**do you have benchmarks for composite video vs vga rendering ?**"

In my project I have extensive use of both Cores of ESP32 and cannot achieve the "VGA_512x384_60Hz" with VGA16Controller in DOUBLEBUFFERING. I am actually limited to a maximum of 314 lines using your option of FABGLIB_MINFREELARGESTBLOCK at 2000!

It would have been better for me to use a video driver like "VGA_480x320_60Hz" if it was available!

I would have created that driver line in "fabglconf.h" if I knew the calculations to have all the values needed... ;)

Will I get better result with your Composite portion of your new library with 16 bits colors and less constraints on the ESP32?

Does it have a DOUBLEBUFFERING as well?

If the PayLoad is less than the VGA on the ESP32 Cores, it would be interesting for me...

Thank you for your answers and also felicitation for your good work.

**Comment by jjplano** [comment]
hi! congrats on the good work! really interested about this feature for analog composite video output

rn I just wanted to ask if the composite output connection is on pin 25 and without any resistors please. I can't seem to find that information anywhere

thanks!

**Comment by ripper121** [comment]
Is this RGB888 Correct?
So its possible to produce 24Bit colors?

**Comment by mobluse** [comment]
Would composite video and VGA be available at the same time from the ESP32? If not, the VGA connector might be used for composite output. I have a TTGO VGA32_1.4. I also have a DuinoMite Mini and to that I have a passive VGA to composite adapter that works. The DuinoMite Mini has no built-in composite connector and only outputs B/W for both VGA and composite to the VGA connector (but not at the same time), see pdf with schematics on https://www.olimex.com/Products/Duino/Duinomite/DUINOMITE-MINI/open-source-hardware. There are a lot of passive VGA to composite video/S-video and audio adapters out there, but they might only work with laptops that are prepared to use them. They might also be used just to use a long VGA cable to a projector to send composite audio and video to the projector. Maybe we can find a passive adapter that works with suitable pins on the VGA connector.

**Comment by Zhornic** [comment]
https://www.drive2.ru/b/561070064489463999/

**Comment by mobluse** [comment]
I tried it by connecting the sound port to a 5.5" B/W CRT TV, but I get a very blurry, but stable picture. When I press Tab on the keyboard there are changes corresponding to the drop down menus. 

Is it blurry because the sound out has a filter? I tried changing MODES_DEFAULT, MONO_DEFAULT, and HRATE_DEFAULT, but it looks equally blurry. I've not tested all combinations. 

I also tried changing VIDEOOUT_GPIO GPIO_NUM_25 to GPIO_NUM_26 and then connecting to hole 5 on the mouse port using a needle, but that didn't work at all. Does GPIO_NUM_26 need to be configured as a D/A port?  Maybe the program could automatically configure the port you choose as a D/A port or warn, during compilation, if not possible. Is it possible to use GPIO_NUM_26/MOUSE HOLE 5 raw or do you need some components connected to it?

I also tried to connect a 5" color LCD monitor with composite in, but that didn't detect a signal; neither from sound out (GPIO_NUM_25) or mouse hole 5 (GPIO_NUM_26).

The CRT TV is a Denver DTV-560 B/W TV with AM/FM radio. The CRT and LCD monitors normally works. The TTGO VGA32_V1.4, e.g. mouse and sound, still works after my experiments.
